{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: dill in /home/elvictor/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages (0.3.3)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install dill"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dill\n",
    "dill.dump_session('updateNotebook.db')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dill\n",
    "dill.load_session(\"updateNotebook.db\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import qiime2\n",
    "from tempfile import mkdtemp\n",
    "from qiime2.plugins import demux, deblur, quality_filter, \\\n",
    "                           metadata, feature_table, alignment, \\\n",
    "                           phylogeny, diversity, emperor, feature_classifier, \\\n",
    "                           taxa, composition\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘qiime2-atacama-tutorial’: File exists\r\n"
     ]
    }
   ],
   "source": [
    "!mkdir qiime2-atacama-tutorial"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "workdir='./'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2020-12-07 19:53:13--  https://data.qiime2.org/2020.8/tutorials/atacama-soils/sample_metadata.tsv\n",
      "Resolving data.qiime2.org (data.qiime2.org)... 52.35.38.247\n",
      "Connecting to data.qiime2.org (data.qiime2.org)|52.35.38.247|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 FOUND\n",
      "Location: https://docs.google.com/spreadsheets/d/1OZtpMzJ0vRhRk_1DA3nYuhA-OLfo6MZYHIHAqRbKVB4/export?gid=0&format=tsv [following]\n",
      "--2020-12-07 19:53:16--  https://docs.google.com/spreadsheets/d/1OZtpMzJ0vRhRk_1DA3nYuhA-OLfo6MZYHIHAqRbKVB4/export?gid=0&format=tsv\n",
      "Resolving docs.google.com (docs.google.com)... 216.58.206.142, 2a00:1450:4009:811::200e\n",
      "Connecting to docs.google.com (docs.google.com)|216.58.206.142|:443... connected.\n",
      "HTTP request sent, awaiting response... 307 Temporary Redirect\n",
      "Location: https://doc-0k-6o-sheets.googleusercontent.com/export/l5l039s6ni5uumqbsj9o11lmdc/11hlo289a1pq6e25rp1n87el9o/1607367195000/103995680502445084602/*/1OZtpMzJ0vRhRk_1DA3nYuhA-OLfo6MZYHIHAqRbKVB4?gid=0&format=tsv [following]\n",
      "Warning: wildcards not supported in HTTP.\n",
      "--2020-12-07 19:53:18--  https://doc-0k-6o-sheets.googleusercontent.com/export/l5l039s6ni5uumqbsj9o11lmdc/11hlo289a1pq6e25rp1n87el9o/1607367195000/103995680502445084602/*/1OZtpMzJ0vRhRk_1DA3nYuhA-OLfo6MZYHIHAqRbKVB4?gid=0&format=tsv\n",
      "Resolving doc-0k-6o-sheets.googleusercontent.com (doc-0k-6o-sheets.googleusercontent.com)... 216.58.213.1, 2a00:1450:4009:816::2001\n",
      "Connecting to doc-0k-6o-sheets.googleusercontent.com (doc-0k-6o-sheets.googleusercontent.com)|216.58.213.1|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: unspecified [text/tab-separated-values]\n",
      "Saving to: ‘sample-metadata.tsv’\n",
      "\n",
      "sample-metadata.tsv     [ <=>                ]   9.21K  --.-KB/s    in 0.04s   \n",
      "\n",
      "2020-12-07 19:53:21 (226 KB/s) - ‘sample-metadata.tsv’ saved [9433]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget \\\n",
    "  -O \"sample-metadata.tsv\" \\\n",
    "  \"https://data.qiime2.org/2020.8/tutorials/atacama-soils/sample_metadata.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir emp-paired-end-sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2020-12-07 20:31:37--  https://data.qiime2.org/2020.8/tutorials/atacama-soils/10p/forward.fastq.gz\n",
      "Resolving data.qiime2.org (data.qiime2.org)... 52.35.38.247\n",
      "Connecting to data.qiime2.org (data.qiime2.org)|52.35.38.247|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 FOUND\n",
      "Location: https://s3-us-west-2.amazonaws.com/qiime2-data/2020.8/tutorials/atacama-soils/10p/forward.fastq.gz [following]\n",
      "--2020-12-07 20:31:39--  https://s3-us-west-2.amazonaws.com/qiime2-data/2020.8/tutorials/atacama-soils/10p/forward.fastq.gz\n",
      "Resolving s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)... 52.218.253.96\n",
      "Connecting to s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)|52.218.253.96|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 143193967 (137M) [binary/octet-stream]\n",
      "Saving to: ‘emp-paired-end-sequences/forward.fastq.gz’\n",
      "\n",
      "emp-paired-end-sequ 100%[===================>] 136.56M   576KB/s    in 5m 47s  \n",
      "\n",
      "2020-12-07 20:37:27 (403 KB/s) - ‘emp-paired-end-sequences/forward.fastq.gz’ saved [143193967/143193967]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget \\\n",
    "  -O \"emp-paired-end-sequences/forward.fastq.gz\" \\\n",
    "  \"https://data.qiime2.org/2020.8/tutorials/atacama-soils/10p/forward.fastq.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2020-12-07 20:28:56--  https://data.qiime2.org/2020.8/tutorials/atacama-soils/10p/reverse.fastq.gz\n",
      "Resolving data.qiime2.org (data.qiime2.org)... 52.35.38.247\n",
      "Connecting to data.qiime2.org (data.qiime2.org)|52.35.38.247|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 FOUND\n",
      "Location: https://s3-us-west-2.amazonaws.com/qiime2-data/2020.8/tutorials/atacama-soils/10p/reverse.fastq.gz [following]\n",
      "--2020-12-07 20:28:57--  https://s3-us-west-2.amazonaws.com/qiime2-data/2020.8/tutorials/atacama-soils/10p/reverse.fastq.gz\n",
      "Resolving s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)... 52.218.241.40\n",
      "Connecting to s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)|52.218.241.40|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 161032441 (154M) [binary/octet-stream]\n",
      "Saving to: ‘emp-paired-end-sequences/reverse.fastq.gz’\n",
      "\n",
      "emp-paired-end-sequ 100%[===================>] 153.57M  1.55MB/s    in 2m 13s  \n",
      "\n",
      "2020-12-07 20:31:12 (1.15 MB/s) - ‘emp-paired-end-sequences/reverse.fastq.gz’ saved [161032441/161032441]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget \\\n",
    "  -O \"emp-paired-end-sequences/reverse.fastq.gz\" \\\n",
    "  \"https://data.qiime2.org/2020.8/tutorials/atacama-soils/10p/reverse.fastq.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2020-12-07 20:27:49--  https://data.qiime2.org/2020.8/tutorials/atacama-soils/10p/barcodes.fastq.gz\n",
      "Resolving data.qiime2.org (data.qiime2.org)... 52.35.38.247\n",
      "Connecting to data.qiime2.org (data.qiime2.org)|52.35.38.247|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 FOUND\n",
      "Location: https://s3-us-west-2.amazonaws.com/qiime2-data/2020.8/tutorials/atacama-soils/10p/barcodes.fastq.gz [following]\n",
      "--2020-12-07 20:27:50--  https://s3-us-west-2.amazonaws.com/qiime2-data/2020.8/tutorials/atacama-soils/10p/barcodes.fastq.gz\n",
      "Resolving s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)... 52.218.250.72\n",
      "Connecting to s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)|52.218.250.72|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 19976093 (19M) [binary/octet-stream]\n",
      "Saving to: ‘emp-paired-end-sequences/barcodes.fastq.gz’\n",
      "\n",
      "emp-paired-end-sequ 100%[===================>]  19.05M  2.27MB/s    in 20s     \n",
      "\n",
      "2020-12-07 20:28:12 (987 KB/s) - ‘emp-paired-end-sequences/barcodes.fastq.gz’ saved [19976093/19976093]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget \\\n",
    "  -O \"emp-paired-end-sequences/barcodes.fastq.gz\" \\\n",
    "  \"https://data.qiime2.org/2020.8/tutorials/atacama-soils/10p/barcodes.fastq.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#IMPORTING DATA AS QIIME2 ATIFACT\n",
    "pair_end_sequences = qiime2.Artifact.import_data('EMPPairedEndSequences', workdir+'/emp-paired-end-sequences')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<artifact: EMPPairedEndSequences uuid: 5cb3b93a-9cd6-4402-aaff-e2f477871205>"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pair_end_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_metadata = qiime2.Metadata.load(workdir+'/sample-metadata.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Demultiplexing processs\n",
    "demux_sequences = demux.methods.emp_paired(pair_end_sequences,\n",
    "                                           sample_metadata.get_column('barcode-sequence'),golay_error_correction =False,rev_comp_mapping_barcodes= True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<artifact: SampleData[PairedEndSequencesWithQuality] uuid: 7d4c5fab-a12f-43dc-87c5-65f534518130>"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demux_sequences.per_sample_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-vjei1ikm')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 6cc73ed5-ffaf-489e-a8f5-3d7e493809a1>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 432x288 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "demux_summary = demux.visualizers.summarize(demux_sequences.per_sample_sequences)\n",
    "demux_summary.visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_paired.R /tmp/tmps6z00d6w/forward /tmp/tmps6z00d6w/reverse /tmp/tmps6z00d6w/output.tsv.biom /tmp/tmps6z00d6w/track.tsv /tmp/tmps6z00d6w/filt_f /tmp/tmps6z00d6w/filt_r 150 150 13 13 2.0 2.0 2 independent consensus 1.0 4 1000000\n",
      "\n"
     ]
    }
   ],
   "source": [
    " # Filter, denoise, pair, and remove chimeras\n",
    "\n",
    "from qiime2.plugins.dada2.methods import denoise_paired\n",
    "\n",
    "clean_seq=denoise_paired(demux_sequences.per_sample_sequences,\n",
    "                     trunc_len_f =150,trunc_len_r =150,\n",
    "                     trim_left_f=13,trim_left_r=13, n_threads=4)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Results (name = value)\n",
       "---------------------------------------------------------------------------------------------------------\n",
       "table                    = <artifact: FeatureTable[Frequency] uuid: 1b9a02a7-5d42-41c8-a5fd-a412d8ffb409>\n",
       "representative_sequences = <artifact: FeatureData[Sequence] uuid: b5a0986e-83e1-4089-8265-833b12389233>\n",
       "denoising_stats          = <artifact: SampleData[DADA2Stats] uuid: 8a7f6dc5-8825-4994-8117-1192d4074860>"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clean_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-qsg0grqu')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: d5b03524-2df0-4eab-91ed-c0a7f4aac6d0>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Visualize representative sequences\n",
    "dada_seq_viz = feature_table.visualizers.tabulate_seqs(clean_seq.representative_sequences)\n",
    "dada_seq_viz.visualization \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/elvictor/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:122: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'sample-frequency-detail.csv'))\n",
      "/home/elvictor/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:124: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'feature-frequency-detail.csv'))\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-vsnfzhfb')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 33968190-fa7c-41d4-ac2e-2b4e51de96ed>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEKCAYAAAD9xUlFAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAWvUlEQVR4nO3de5CldX3n8fcHEEWFMQqmlIuDDkHZ0hjTysa1Es1GA1oDgoiCuWhYBiwxurta4opi1t21rFw2ywbRSULG7AYIEEOYiMFLgkQXcQYQw0WKWVZgghtG3QwXowT47h/nmcOhPd39dE8//fTpfr+qTs15fue5fPvHoT/93H5PqgpJkgD26LsASdLyYShIkoYMBUnSkKEgSRoyFCRJQ4aCJGlor74L2B37779/rV27tu8yJGmiXHfddd+pqgPGfTbRobB27Vq2bt3adxmSNFGS3DnTZx4+kiQNTWQoJFmfZOPOnTv7LkWSVpSJDIWq2lxVG9asWdN3KZK0okxkKEiSumEoSJKGDAVJ0pChIEkamuj7FHbHBdfe1XcJEgAnH3lI3yVIQ+4pSJKGDAVJ0pChIEkaMhQkSUMTGQoOcyFJ3ZjIUHCYC0nqxkSGgiSpG4aCJGnIUJAkDRkKkqQhQ0GSNGQoSJKGDAVJ0pChIEkaMhQkSUOGgiRpyFCQJA0ZCpKkoWUTCklekOQTSS5N8va+65Gk1ajTUEhyfpJ7k9w0rf2oJLcl2ZbkTICqurWqTgdOBKa6rEuSNF7XewqbgKNGG5LsCZwLHA0cAZyU5Ijms2OALwNf7LguSdIYnYZCVV0NfG9a88uAbVV1R1U9BFwEHNvMf3lVvRx4y0zrTLIhydYkW3fs2NFV6ZK0Ku3VwzYPBO4emd4OHJnklcDxwBOBK2ZauKo2AhsBpqamqrsyJWn16SMUMqatquoq4KqlLUWSNKqPq4+2AwePTB8E3DOfFfiMZknqRh+hsAU4LMmhSfYG3gxcPp8V+IxmSepG15ekXghcAxyeZHuSU6rqYeAM4ErgVuDiqrq5yzokSe10ek6hqk6aof0KZjmZPJck64H169atW+gqJEljLJs7mufDw0eS1I2JDAVJUjcMBUnS0ESGgpekSlI3JjIUPKcgSd2YyFCQJHVjIkPBw0eS1I2JDAUPH0lSNyYyFCRJ3TAUJElDhoIkaWgiQ8ETzZLUjYkMBU80S1I3JjIUJEndMBQkSUOGgiRpyFCQJA1NZCh49ZEkdWMiQ8GrjySpGxMZCpKkbhgKkqQhQ0GSNGQoSJKGDAVJ0tCcoZDkjUn2bd6fleTTSV7SfWmz1uQlqZLUgTZ7Ch+sqvuTvAL4ReBTwHndljU7L0mVpG60CYVHmn9fB5xXVX8B7N1dSZKkvrQJhb9P8kngROCKJE9suZwkacK0+eV+InAlcFRV/SPwdOC9nVYlSerFXnPNUFXfT3Iv8ArgduDh5l9Ji+CCa+/qfBsnH3lI59vQytDm6qOzgfcB72+angD8zy6LkiT1o83ho+OAY4AHAarqHmDfLouSJPWjTSg8VFUFFECSp3RbkiSpL21C4eLm6qOnJTkV+ALw+92WJUnqw6wnmpME+FPg+cB9wOHAh6rq80tQ22x1rQfWr1u3rs8yJGnFmTUUqqqSXFZVPw30GgSjqmozsHlqaurUvmuRpJWkzeGjryZ5aeeVSJJ6N+d9CsCrgNOS3MngCqQw2Il4UaeVSZKWXJtQOLrzKiRJy0KbUKjOq5AkLQttQuEzDIIhwJOAQ4HbgH/RYV2SpB60GfvohaPTzQN2TuusIklSb+Y9BHZVXQ94NZIkrUBz7ikk+Xcjk3sALwF2dFaRJKk3bc4pjA5+9zCDcwx/1k05kqQ+tQmFW6rqktGGJG8ELplhfknShGpzTuH9LdskSRNuxj2FJEcDrwUOTHLOyEf7MTiMJElaYWY7fHQPsJXBA3auG2m/H/i3XRYlSerHjKFQVTcCNya5oKr+eSmKSfJ64HXAM4Fzq+pzS7FdSdJAm3MKa5NcmuSWJHfserXdQJLzk9yb5KZp7UcluS3JtiRnAlTVZVV1KvBW4E3z+UEkSbuvzdVHfwScDfxXBiOmvo3BkBdtbQJ+D/jjXQ1J9gTOBV4NbAe2JLm8qm5pZjmr+VzSIrjg2rsWvOzJRx6yiJVouWuzp7BPVX0RSFXdWVUfBn6+7Qaq6mrge9OaXwZsq6o7quoh4CLg2Ax8DPhsc+f0j0iyIcnWJFt37PAeOklaTG1C4QdJ9gBuT3JGkuMYHPPfHQcCd49Mb2/a3gn8AnBCktPHLVhVG6tqqqqmDjjggN0sQ5I0qs3ho3cDTwZ+HfgIg0NIv7qb2x13+Kmq6hzgnDGfSZKWQJtRUrcAJKmqetsibXc7cPDI9EEMLoFtJcl6YP26desWqRxJErQ4fJTkZ5LcAtzaTP9kko/v5na3AIclOTTJ3sCbgcvbLlxVm6tqw5o1a3azDEnSqDbnFH4X+EXguzC8f+Fn224gyYXANcDhSbYnOaWqHgbOAK5kEDYXV9XN8y1ekrS42pxToKruTh53GuCRthuoqpNmaL8CuKLtekZ5+EiSutFmT+HuJC8HKsneSd5DcyipLx4+kqRutAmF04F3MLhkdDvw4mZakrTCzDZK6seq6n3Aq6rqLUtY05w8fCRJ3ZhtT+G1SZ7AMnx2goePJKkbs51o/ivgO8BTktzH4Iaz2vVvVe23BPVJkpbQjHsKVfXeqloDfKaq9quqfUf/XcIaJUlLZM4TzVV17FIUMh9J1ifZuHPnzr5LkaQVpc3VR8uO5xQkqRsTGQqSpG7MGApJvtj8+7GlK0eS1KfZrj56VpKfA45JchHThrue6SE4kqTJNVsofAg4k8Gw1r8z7bNiHk9fW2zevCZJ3ZgxFKrqUuDSJB+sqo8sYU1zqqrNwOapqalT+65FklaSNg/Z+UiSY3hsuOyrquovuy1L0nJxwbV3tZrv5CMP6bgSLYU2D9n5KPAu4Jbm9a6mTZK0wrR5nsLrgBdX1aMAST4F3MAyHBNJkrR72t6n8LSR973fMeYdzZLUjTah8FHghiSbmr2E64D/0m1Zs/OOZknqRpsTzRcmuQp4KYN7Fd5XVf+368IkSUuv7TOavw1c3nEtkqSeOfaRJGnIUJAkDc0aCkn2SHLTUhUjSerXrKHQ3JtwYxJvVZSkVaDNieZnATcn+Rrw4K7Gqjqms6rm4IB4ktSNNqHwG51XMU8OiCctP3ONkeTYSJOhzX0KX0ryHOCwqvpCkicDe3ZfmiRpqbUZEO9U4FLgk03TgcBlXRYlSepHm0tS3wH8K+A+gKq6HXhml0VJkvrRJhR+WFUP7ZpIsheDJ69JklaYNqHwpST/AdgnyauBS4DN3ZYlSepDm1A4E9gB/B1wGnAFcFaXRUmS+tHm6qNHmyGzr2Vw2Oi2qvLwkSStQHOGQpLXAZ8A/jeDobMPTXJaVX226+IkSUurzc1rvw28qqq2ASR5HvAZoLdQ8I5mSepGm3MK9+4KhMYdwL0d1dOKT16TpG7MuKeQ5Pjm7c1JrgAuZnBO4Y3AliWoTZK0xGY7fLR+5P0/AD/XvN8B/FhnFUlakWYbG8lxkZaPGUOhqt62lIVIWr3GBYZB0Y82Vx8dCrwTWDs6f59DZ0uSutHm6qPLgD9kcBfzo92WI0nqU5tQ+EFVndN5JZKk3rUJhf+W5Gzgc8APdzVW1fWdVSVJ6kWbUHgh8MvAz/PY4aNqpiVJK0ibUDgOeO7o8NmSpJWpzR3NNwJP67oQSVL/2uwp/DjwzSRbePw5BS9JlaQVpk0onN15FZKkZaHN8xS+tBSFJHku8AFgTVWdsBTblCQ93pznFJLcn+S+5vWDJI8kua/NypOcn+TeJDdNaz8qyW1JtiU5E6Cq7qiqUxb2Y0iSFsOcoVBV+1bVfs3rScAbgN9ruf5NwFGjDUn2BM4FjgaOAE5KcsS8qpYkdaLN1UePU1WX0fIehaq6GvjetOaXAduaPYOHgIuAY+dbhyRp8bUZEO/4kck9gCkGN68t1IHA3SPT24EjkzwD+M/ATyV5f1V9dIZ6NgAbAA45xFEUJWkxtbn6aPS5Cg8D32L3/rLPmLaqqu8Cp8+1cFVtBDYCTE1N7U44SZKmaXP10WI/V2E7cPDI9EHAPfNZgc9ollY+n7HQj9kex/mhWZarqvrIAre5BTiseU7D3wNvBk6ezwqqajOweWpq6tQF1iBJGmO2E80PjnkBnAK8r83Kk1wIXAMcnmR7klOq6mHgDOBK4Fbg4qq6eYH1S5IWUarmPiyfZF/gXQwC4WLgt6vq3o5rm62eXYePTr399tsXtI7ZnhcraTJ4OGlhklxXVVPjPpv1ktQkT0/yn4BvMDjU9JKqel+fgQCDw0dVtWHNmjV9liFJK85s5xR+EziewZU+L6yqB5asKklSL2bbU/j3wLOBs4B7Roa6uL/tMBeSpMky455CVc37buel4iWpktSNZfuLfzaeU5CkbkxkKEiSumEoSJKG2ox9tOx4TkES/Oj9Rt63sPsmck/BcwqS1I2JDAVJUjcMBUnSkOcUJK1Yo+ccPN/QzkTuKXhOQZK6MZGhIEnqhqEgSRoyFCRJQ4aCJGloIkMhyfokG3fu3Nl3KZK0okxkKHj1kSR1YyJDQZLUDUNBkjRkKEiShgwFSdKQoSBJGprIUPCSVEnzdcG1dw1f811mrnlWkokMBS9JlaRuTGQoSJK6YShIkoYMBUnSkKEgSRoyFCRJQ4aCJGnIUJAkDRkKkqShvfouYCGSrAfWr1u3ru9SJC0jbe8uHjffyUceMq9lR+cf1zZ9mbbrb1vDYq5v1ETuKXhHsyR1YyJDQZLUDUNBkjRkKEiShgwFSdKQoSBJGjIUJElDhoIkachQkCQNGQqSpCFDQZI0ZChIkoYMBUnSkKEgSRpaNkNnJ3kK8HHgIeCqqvqTnkuSpFWn0z2FJOcnuTfJTdPaj0pyW5JtSc5smo8HLq2qU4FjuqxLkjRe14ePNgFHjTYk2RM4FzgaOAI4KckRwEHA3c1sj3RclyRpjE5DoaquBr43rfllwLaquqOqHgIuAo4FtjMIhlnrSrIhydYkW3fs2NFF2ZJWoQuuvWvOJ7eNfj5u/nHTbZ4G12Zdc7Uvlj5ONB/IY3sEMAiDA4FPA29Ich6weaaFq2pjVU1V1dQBBxzQbaWStMr0caI5Y9qqqh4E3rbUxUiSHtPHnsJ24OCR6YOAe+azgiTrk2zcuXPnohYmSatdH6GwBTgsyaFJ9gbeDFw+nxVU1eaq2rBmzZpOCpSk1arrS1IvBK4BDk+yPckpVfUwcAZwJXArcHFV3dxlHZKkdjo9p1BVJ83QfgVwxULXm2Q9sH7dunULXYUkaYyJHObCw0eS1I2JDAVJUjcmMhS8+kiSupGq6ruGBUuyA7gTWANMT4jpbdOn9we+02mB4+vqYtm55p3p8/m0r5b+XGhfzvTZau7LNvP63eznu/mcqhp/929VTfwL2DhX25jprX3U1cWyc8070+fzaV8t/bnQvmzbb6upL3enP/1uLl5fzre+iTx8NMa4YTGmt804dEaHdmeb81l2rnln+nw+7aulPxfalzN9tpr7ss28fjf7+26ONdGHj3ZHkq1VNdV3HSuF/bl47MvFZX/Oz0rZU1iIjX0XsMLYn4vHvlxc9uc8rNo9BUnSj1rNewqSpGkMBUnSkKEgSRoyFBpJnpLkU0l+P8lb+q5nkiV5bpI/THJp37WsBEle33wv/yLJa/quZ5IleUGSTyS5NMnb+65nOVrRoZDk/CT3JrlpWvtRSW5Lsi3JmU3z8cClVXUqcMySF7vMzacva/D87VP6qXQyzLM/L2u+l28F3tRDucvaPPvy1qo6HTgR8DLVMVZ0KACbgKNGG5LsCZwLHA0cAZyU5AgGT4Db9ezoR5awxkmxifZ9qbltYv79eVbzuR5vE/PoyyTHAF8Gvri0ZU6GFR0KVXU18L1pzS8DtjV/zT4EXAQcy+AxoQc186zoflmIefal5jCf/szAx4DPVtX1S13rcjff72ZVXV5VLwc8TDzGavzldyCP7RHAIAwOBD4NvCHJefRzm/wkGtuXSZ6R5BPATyV5fz+lTaSZvpvvBH4BOCHJ6X0UNoFm+m6+Msk5ST7JbjzoayXr9Mlry1TGtFVVPQi8bamLmXAz9eV3AX95zd9M/XkOcM5SFzPhZurLq4CrlraUybIa9xS2AwePTB8E3NNTLZPOvlxc9ufisS8XaDWGwhbgsCSHJtkbeDNwec81TSr7cnHZn4vHvlygFR0KSS4ErgEOT7I9ySlV9TBwBnAlcCtwcVXd3Gedk8C+XFz25+KxLxeXA+JJkoZW9J6CJGl+DAVJ0pChIEkaMhQkSUOGgiRpyFCQJA0ZCupVkkeSfH3ktbbvmpajJM9v+ueGJM9bwPLvTvLkLmrTyuJ9CupVkgeq6qkzfBYG39FHl7is3iXZs6oeGZk+E9inqs5e4Pq+BUxV1XfmscxezU1gWkXcU9CykmRtkluTfBy4Hjg4yXuTbEnyjSS/MTLvB5qHqHwhyYVJ3tO0X5Vkqnm/f/MLkSR7JvnNkXWd1rS/slnm0iTfTPInTSCR5KVJ/leSG5N8Lcm+Sf42yYtH6vhKkhdN+znemsGT0v6qqfHskc9+qVnX15N8shn7nyQPJPmPSa4FfmZk/tcC7wb+TZK/mWMd5yXZmuTmXX2V5NeBZwN/M7L8AyPrPyHJpub9piS/08z3sSTPa36G65qf+/kL/6+riVBVvnz19mLwQKOvN68/B9YCjwL/svn8NcBGBqNe7gH8JfCzwE8Dfwc8GdgP2Aa8p1nmKgZ/FQPsD3yreb8BOKt5/0RgK3Ao8EpgJ4NB0/ZgMGTCK4C9gTuAlzbL7MdgZOFfBX63afsJYOuYn+utwLeBZwD7ADcxeNLXCxgMzf6EZr6PA7/SvC/gxBn66cMjP99s63h68++eTT+8qJn+FrD/yPoeGHl/ArCpeb+p6eM9m+kvAoc1748E/rrv74yvbl+rcehsLS//VFWjf3WvBe6sqq82Ta9pXjc0008FDgP2Bf68qr7fLNdmsLPXAC9KckIzvaZZ10PA16pqe7OurzMIp53At6tqC0BV3dd8fgnwwSTvBX6NwS/ScT5fg2HESfJpBkHzMINA29LsjOwD3NvM/wjwZy1+jn89yzpOTLKBQXg9i8FTx77RYp2jLqmqR5I8FXg5cEmzHRiEqVYwQ0HL0YMj7wN8tKo+OTpDkncz+Mt6nId57NDok6at651VdeW0db0S+OFI0yMM/t/IuG1U1feTfJ7Bk7xme9bv9GWrWeenqmrcw4d+UCPnEWYxdh1JDgXew2DP5v81h4SeNGb56bVNn2dX/+8B/ONoaGvl85yClrsrgV9r/molyYFJnglcDRyXZJ8k+wLrR5b5FoO/pGFwaGR0XW9P8oRmXT+R5CmzbPubwLOTvLSZf98ku/6Q+gMGD77ZUlXTHwW5y6uTPD3JPsDrga8wOBxzQvMz0Hz+nLm74XFmWsd+DH6h70zy4wyeT7zL/Qz2rnb5hyQvSLIHcNy4jTR7Rv8nyRub7STJT86zVk0Y9xS0rFXV55K8ALimOYTxAPBLVXV9kj9lcC7iTuBvRxb7LeDiJL8M/PVI+x8wOCx0fXMieQeDX9YzbfuhJG8C/nvzi/2fGDwW84Gqui7JfcAfzVL+l4H/AawDLqiqrQBJzgI+1/xC/mfgHc3P0EpV3TJuHVX11SQ3ADczOBfylZHFNgKfTfLtqnoVcCaDcwd3MzjfMfYKMAbPMT6v2d4TGDzr+Ma2tWryeEmqVoQkH2bwy/q3lmh7z2ZwIvf5NeaS2SRvZXCy+4ylqEdaLB4+kuYpya8A1wIfGBcI0iRzT0GSNOSegiRpyFCQJA0ZCpKkIUNBkjRkKEiShgwFSdLQ/weIqv7tUos7CgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Visualize feature table\n",
    "\n",
    "feature_table_viz = feature_table.visualizers.summarize(clean_seq.table,\n",
    "                                                        sample_metadata)\n",
    "feature_table_viz.visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Taxonomic analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2020-12-07 21:03:25--  https://data.qiime2.org/2020.8/common/gg-13-8-99-515-806-nb-classifier.qza\n",
      "Resolving data.qiime2.org (data.qiime2.org)... 52.35.38.247\n",
      "Connecting to data.qiime2.org (data.qiime2.org)|52.35.38.247|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 FOUND\n",
      "Location: https://s3-us-west-2.amazonaws.com/qiime2-data/2020.8/common/gg-13-8-99-515-806-nb-classifier.qza [following]\n",
      "--2020-12-07 21:03:27--  https://s3-us-west-2.amazonaws.com/qiime2-data/2020.8/common/gg-13-8-99-515-806-nb-classifier.qza\n",
      "Resolving s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)... 52.218.136.64\n",
      "Connecting to s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)|52.218.136.64|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 28285585 (27M) [binary/octet-stream]\n",
      "Saving to: ‘gg-13-8-99-515-806-nb-classifier.qza’\n",
      "\n",
      "gg-13-8-99-515-806- 100%[===================>]  26.97M  1.36MB/s    in 35s     \n",
      "\n",
      "2020-12-07 21:04:02 (799 KB/s) - ‘gg-13-8-99-515-806-nb-classifier.qza’ saved [28285585/28285585]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#https://pubmed.ncbi.nlm.nih.gov/28593197/\n",
    "\n",
    "!wget \\\n",
    "  -O \"gg-13-8-99-515-806-nb-classifier.qza\" \\\n",
    "  \"https://data.qiime2.org/2020.8/common/gg-13-8-99-515-806-nb-classifier.qza\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mExported .//gg-13-8-99-515-806-nb-classifier.qza as TaxonomicClassiferTemporaryPickleDirFmt to directory .//classifier\u001b[0m\r\n"
     ]
    }
   ],
   "source": [
    "!qiime tools export --input-path $workdir/gg-13-8-99-515-806-nb-classifier.qza --output-path $workdir/classifier\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "gg_classifier = qiime2.Artifact.import_data('TaxonomicClassifier', workdir+'/classifier/')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy = feature_classifier.methods.classify_sklearn(reads = clean_seq.representative_sequences,\n",
    "                                                       classifier = gg_classifier)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-0ujrcowo')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: c24a973a-1ffc-4503-aa94-f0c95c47b93a>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonomy_classification = metadata.visualizers.tabulate(taxonomy.classification.view(qiime2.Metadata))\n",
    "taxonomy_classification.visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-80o3vfq3')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 152db80a-2b6a-4038-9a4b-a1ae6a5f4464>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxa_bar_plot = taxa.visualizers.barplot(clean_seq.table, taxonomy.classification, sample_metadata)\n",
    "taxa_bar_plot.visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differential abundance testing with ANCOM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Differential abundance testing with ANCOM¶\n",
    "BAQ1370_site = feature_table.methods.filter_samples(clean_seq.table,\n",
    "                                                  metadata = sample_metadata,\n",
    "                                                  where = \"siteName='BAQ1370'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "All values in `grouping` are unique. This method cannot operate on a grouping vector with only unique values (e.g., there are no 'within' variance because each group of samples contains only a single sample).",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-30-0c66982f0b35>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m ancom_BAQ1370_site = composition.actions.ancom(table = BAQ1370_site_composition.composition_table,\n\u001b[0;32m----> 4\u001b[0;31m                                              metadata = sample_metadata.get_column('transectName'))\n\u001b[0m",
      "\u001b[0;32m<decorator-gen-537>\u001b[0m in \u001b[0;36mancom\u001b[0;34m(table, metadata, transform_function, difference_function)\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/qiime2/sdk/action.py\u001b[0m in \u001b[0;36mbound_callable\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    243\u001b[0m                 \u001b[0;31m# Execute\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    244\u001b[0m                 outputs = self._callable_executor_(scope, callable_args,\n\u001b[0;32m--> 245\u001b[0;31m                                                    output_types, provenance)\n\u001b[0m\u001b[1;32m    246\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    247\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutputs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msignature\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moutputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/qiime2/sdk/action.py\u001b[0m in \u001b[0;36m_callable_executor_\u001b[0;34m(self, scope, view_args, output_types, provenance)\u001b[0m\n\u001b[1;32m    450\u001b[0m         \u001b[0;31m# will also need to be updated to support OutPath instead of str.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    451\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mtempfile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTemporaryDirectory\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprefix\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'qiime2-temp-'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mtemp_dir\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 452\u001b[0;31m             \u001b[0mret_val\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_callable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput_dir\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtemp_dir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mview_args\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    453\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mret_val\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    454\u001b[0m                 raise TypeError(\n",
      "\u001b[0;32m~/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/q2_composition/_ancom.py\u001b[0m in \u001b[0;36mancom\u001b[0;34m(output_dir, table, metadata, transform_function, difference_function)\u001b[0m\n\u001b[1;32m     54\u001b[0m     ancom_results = skbio_ancom(table,\n\u001b[1;32m     55\u001b[0m                                 \u001b[0mmetadata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_series\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 56\u001b[0;31m                                 significance_test=f_oneway)\n\u001b[0m\u001b[1;32m     57\u001b[0m     \u001b[0mancom_results\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mby\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'W'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mascending\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m     ancom_results[0].rename(columns={'reject': 'Reject null hypothesis'},\n",
      "\u001b[0;32m~/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/skbio/stats/composition.py\u001b[0m in \u001b[0;36mancom\u001b[0;34m(table, grouping, alpha, tau, theta, multiple_comparisons_correction, significance_test, percentiles)\u001b[0m\n\u001b[1;32m   1079\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mnum_groups\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgrouping\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1080\u001b[0m         raise ValueError(\n\u001b[0;32m-> 1081\u001b[0;31m             \u001b[0;34m\"All values in `grouping` are unique. This method cannot \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1082\u001b[0m             \u001b[0;34m\"operate on a grouping vector with only unique values (e.g., \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1083\u001b[0m             \u001b[0;34m\"there are no 'within' variance because each group of samples \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: All values in `grouping` are unique. This method cannot operate on a grouping vector with only unique values (e.g., there are no 'within' variance because each group of samples contains only a single sample)."
     ]
    }
   ],
   "source": [
    "BAQ1370_site_composition = composition.actions.add_pseudocount(BAQ1370_site.filtered_table)\n",
    "\n",
    "ancom_BAQ1370_site = composition.actions.ancom(table = BAQ1370_site_composition.composition_table,\n",
    "                                             metadata = sample_metadata.get_column('transectName'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# #Generate a tree for phylogenetic diversity analyses¶"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running external command line application. This may print messages to stdout and/or stderr.\n",
      "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: mafft --preservecase --inputorder --thread -1 /tmp/qiime2-archive-dsakr9v6/b5a0986e-83e1-4089-8265-833b12389233/data/dna-sequences.fasta\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Generate a tree for phylogenetic diversity analyses¶\n",
    "#First, we perform a multiple sequence alignment of the sequences in our FeatureData[Sequence] \n",
    "#to create a FeatureData[AlignedSequence] QIIME 2 artifact. Here we do this with the mafft program\n",
    "mafft_alignment = alignment.methods.mafft(clean_seq.representative_sequences, n_threads='auto')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Next, we mask (or filter) the alignment to remove positions that are highly variable. \n",
    "#These positions are generally considered to add noise to a resulting phylogenetic tree.\n",
    "masked_mafft_alignment = alignment.methods.mask(mafft_alignment.alignment)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running external command line application. This may print messages to stdout and/or stderr.\n",
      "The command being run is below. This command cannot be manually re-run as it will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: FastTree -quote -nt /tmp/qiime2-archive-q_jzr_k5/4bbe21e3-2d45-4cdd-8f27-3eda9b9bf12d/data/aligned-dna-sequences.fasta\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Next, we’ll apply FastTree to generate a phylogenetic tree from the masked alignment.\n",
    "unrooted_tree = phylogeny.methods.fasttree(masked_mafft_alignment.masked_alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The FastTree program creates an unrooted tree, so in the final step in this section \n",
    "#we apply midpoint rooting to place the root of the tree at the midpoint of the longest tip-to-tip distance in the unrooted tree.\n",
    "rooted_tree = phylogeny.methods.midpoint_root(unrooted_tree.tree)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alpha and beta diversity analysis¶\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/elvictor/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/sklearn/metrics/pairwise.py:1761: DataConversionWarning: Data was converted to boolean for metric jaccard\n",
      "  warnings.warn(msg, DataConversionWarning)\n",
      "/home/elvictor/anaconda3/envs/qiime2-2020.8/lib/python3.6/site-packages/skbio/stats/ordination/_principal_coordinate_analysis.py:152: RuntimeWarning: The result contains negative eigenvalues. Please compare their magnitude with the magnitude of some of the largest positive eigenvalues. If the negative ones are smaller, it's probably safe to ignore them, but if they are large in magnitude, the results won't be useful. See the Notes section for more details. The smallest eigenvalue is -0.011284496461492186 and the largest is 0.8807872254293366.\n",
      "  RuntimeWarning\n"
     ]
    }
   ],
   "source": [
    "core_metrics = diversity.pipelines.core_metrics_phylogenetic(table = clean_seq.table,\n",
    "                                                             phylogeny = rooted_tree.rooted_tree,\n",
    "                                                             sampling_depth = 1109,\n",
    "                                                             metadata = sample_metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# alpha group significance¶\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "faith_pd_group_significance = diversity.actions.alpha_group_significance(core_metrics.faith_pd_vector,\n",
    "                                                                         sample_metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "evenness_group_significance = diversity.actions.alpha_group_significance(core_metrics.evenness_vector,\n",
    "                                                                         sample_metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-mxlwpayk')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 9e4cf73c-08e6-43af-b4c1-0cb1404ec046>"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "faith_pd_group_significance.visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-26vgtxfz')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: c0d60710-2ef0-4d45-a2e1-38a1a0efcbcc>"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "evenness_group_significance.visualization"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# beta group significance¶\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 432x288 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "uUniFrac_siteName_significance = diversity.actions.beta_group_significance(core_metrics.unweighted_unifrac_distance_matrix,\n",
    "                                                                           sample_metadata.get_column('siteName'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-fnp0xjhn')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 13cd00a0-f3e5-478c-bd32-126aca9f0ced>"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uUniFrac_siteName_significance.visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
